src/hooks/useWalletsHiddenBalances.ts

```diff
@@ -0,0 +1,113 @@
+import { useConnectedToHardhatStore } from '@/state/connectedToHardhat';
+import { NativeCurrencyKey } from '@/entities/nativeCurrencyTypes';
+import { userAssetsStore } from '@/state/assets/userAssets';
+import { queryClient } from '@/react-query';
+import { userAssetsQueryKey, UserAssetsResult } from '@/resources/assets/UserAssetsQuery';
+import { convertAmountAndPriceToNativeDisplay, add, isEqual } from '@/helpers/utilities';
+import { isEqual as _isEqual } from 'lodash';
+
+export type WalletBalanceResult = {
+  hiddenBalances: Record<Address, string>;
+};
+
+const getHiddenAssetBalance = ({
+  address,
+  nativeCurrency,
+  connectedToHardhat,
+}: {
+  address: Address;
+  nativeCurrency: NativeCurrencyKey;
+  connectedToHardhat: boolean;
+}) => {
+  const hiddenAssetIds = userAssetsStore.getState(address).getHiddenAssetsIds();
+  const assetData = queryClient.getQueryData<UserAssetsResult>(
+    userAssetsQueryKey({ address, currency: nativeCurrency, connectedToHardhat })
+  );
+
+  const balance = hiddenAssetIds.reduce((acc, uniqueId) => {
+    const asset = assetData?.[uniqueId];
+    let sum = acc;
+    if (asset) {
+      const priceUnit = asset.price?.value ?? 0;
+      const nativeDisplay = convertAmountAndPriceToNativeDisplay(asset?.balance?.amount ?? 0, priceUnit, nativeCurrency);
+      sum = add(sum, nativeDisplay.amount);
+    }
+    return sum;
+  }, '0');
+
+  return balance;
+};
+
+const useWalletsHiddenBalances = (wallets: AllRainbowWallets): WalletBalanceResult => {
+  const { nativeCurrency } = useAccountSettings();
+  const connectedToHardhat = useConnectedToHardhatStore(state => state.connectedToHardhat);
+  const [hiddenBalances, setHiddenBalances] = useState<Record<Address, string>>({});
+
+  const allAddresses = useMemo(
+    () => Object.values(wallets).flatMap(wallet => (wallet.addresses || []).map(account => account.address as Address)),
+    [wallets]
+  );
+
+  const calculateHiddenBalanceForAddress = useCallback(
+    (address: Address) => {
+      const lowerCaseAddress = address.toLowerCase() as Address;
+      const hiddenAssetBalance = getHiddenAssetBalance({ address, nativeCurrency, connectedToHardhat });
+
+      setHiddenBalances(prev => {
+        const newBalance = hiddenAssetBalance;
+        if (!prev[lowerCaseAddress] || !isEqual(prev[lowerCaseAddress], newBalance)) {
+          return {
+            ...prev,
+            [lowerCaseAddress]: newBalance,
+          };
+        }
+        return prev;
+      });
+    },
+    [nativeCurrency, connectedToHardhat]
+  );
+
+  useEffect(() => {
+    allAddresses.forEach(address => {
+      calculateHiddenBalanceForAddress(address);
+    });
+  }, [allAddresses, calculateHiddenBalanceForAddress]);
+
+  useEffect(() => {
+    const unsubscribeFromQueryCache = queryClient.getQueryCache().subscribe(event => {
+      const [args, key] = event.query.queryKey;
+      if (key === 'userAssets') {
+        calculateHiddenBalanceForAddress(args.address);
```

Comment:
do we need to check if this address is in `allAddresses`? not sure it will ever not be the case but